pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,312 @@
+	epDataStore map[endpointID]*endpointMetadata
+}
+
+func NewEgressPolicyManager() *Manager {
+	return &Manager{
+		policyEndpoints: make(map[endpointID][]policyID),
+		policyConfigs:   make(map[policyID]*EgressPolicyConfig),
+		epDataStore:     make(map[endpointID]*endpointMetadata),
+	}
+}
+
+// Event handlers
+
+// AddEgressPolicy parses the given policy config, and updates internal state with the config fields.
+// returns bool indicates if policy is added, err inidates first encountered error
+func (manager *Manager) AddEgressPolicy(config EgressPolicyConfig) (bool, error) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	_, ok := manager.policyConfigs[config.id]
+	if ok {
+		log.WithField(
+			logfields.CiliumEgressNATPolicyName, config.id.Name).Warn("CiliumEgressNATPolicy already exists and is not re-added")
+		return false, errors.New("already exists")
+	}
+
+	err := manager.isValidConfig(config)
+	if err != nil {
+		return false, err
+	}
+
+	manager.policyConfigs[config.id] = &config
+	for _, endpoint := range manager.epDataStore {
+		if config.policyConfigSelectsEndpoint(endpoint) {
+			if err := manager.upsertPolicyEndpoint(&config, endpoint); err != nil {
+				return false, err
+			}
+		}
+	}
+
+	return true, nil
+}
+
+// Deletes the internal state associated with the given policy, including egress eBPF map entries
+func (manager *Manager) DeleteEgressPolicy(configID policyID) error {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	storedConfig := manager.policyConfigs[configID]
+	if storedConfig == nil {
+		return fmt.Errorf("policy not found")
+	}
+	log.WithFields(logrus.Fields{"policyID": configID}).
+		Debug("Delete local egress policy")
+
+	for endpointId, policies := range manager.policyEndpoints {
+		var newPolicyList []policyID
+		// make a new list excluding policy that is to be deleted
+		for _, policyId := range policies {
+			if policyId == storedConfig.id {
+				// found policy to endpoint mapping, need to delete egress map entry
+				// identified by endpoint and config
+				epData, ok := manager.epDataStore[endpointId]
+				if !ok {
+					return fmt.Errorf("failed to get endpoint data for %v", endpointId)
+				}
+				if err := manager.deleteEgressMap(storedConfig, epData); err != nil {
+					return err
+				}
+			} else {
+				newPolicyList = append(newPolicyList, policyId)
+			}
+		}
+		if len(newPolicyList) > 0 {
+			manager.policyEndpoints[endpointId] = newPolicyList
+		} else {
+			// epDataStore untouched here since endpoint data is unchanged
+			delete(manager.policyEndpoints, endpointId)
+		}
+	}
+	delete(manager.policyConfigs, configID)
+	return nil
+}
+
+// OnUpdateEndpoint is the event handler for endpoint additions and updates.
+func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	var epData *endpointMetadata
+	var err error
+
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	if len(endpoint.Networking.Addressing) == 0 {
```

Comment:
It's a good idea to check if `endpoint.Networking` is nil. I'm not sure if it can, but it's better to check it on the safe side.

Reply:
This is a good point. However the place where it is called is already under the condition that "if endpoint.Networking != nil " so this is fine. That case do exist if I did not have the guard early, and would result in a nil pointer deref.